### PR TITLE
feat: add contract upgrade cli

### PR DESCRIPTION
**Pull request by PainterPuppets**
```bash
Usage: kuai contract upgrade [options]

Options:
  --name [string]            name of the contract to be deployed (default: "")
  --bin-path [path]          path of contract bin file (default: "")
  --migration-path <string>  path of migration file
  --feepayer <string...>     address or multisig config of the transaction fee payer
                             (default: "")
  --deployer <string...>     address or multisig config of the contract deployer
                             (default: "")
  --signer [string]          signer provider [default: ckb-cli] [possible values:
                             ckb-cli, ckb-cli-multisig] (default: "")
  --fee-rate [number]        per transaction's fee, deployment may involve more than
                             one transaction. default: [1000] shannons/Byte
                             (default: 1000)
  --export [path]            export transaction to file (default: "")
  --send                     send transaction directly (default: false)
```

### CODE REVIEW

File: packages/core/src/builtin-tasks/contract.ts [hunk 1]
```diff
@@ -16,6 +16,18 @@ import { task, subtask } from '../config/config-env'
 import { paramTypes } from '../params'
 import { getUserConfigPath } from '../project-structure'
 import { getGenesisScriptsConfig } from '../util/chain'
+import { Hash } from '@ckb-lumos/base'
+
+type ContractMigration = {
+  cell_recipes: Array<{
```
Comment (Keith-CY) [inline packages/core/src/builtin-tasks/contract.ts:22]: What does `recipe` mean in `migration`

File: packages/core/src/builtin-tasks/contract.ts [hunk 2]
```diff
@@ -16,6 +16,18 @@ import { task, subtask } from '../config/config-env'
 import { paramTypes } from '../params'
 import { getUserConfigPath } from '../project-structure'
 import { getGenesisScriptsConfig } from '../util/chain'
+import { Hash } from '@ckb-lumos/base'
```
Comment (Keith-CY) [inline packages/core/src/builtin-tasks/contract.ts:19]:
```suggestion
import type { Hash } from '@ckb-lumos/base'
```

File: packages/core/src/builtin-tasks/contract.ts [hunk 3]
```diff
@@ -198,6 +210,181 @@ subtask('contract:deploy')
     return tx
   })
 
+interface ContractUpgradeArgs {
+  name?: string
+  binPath?: string
+  migrationPath: string
+  signer?: 'ckb-cli' | 'ckb-cli-multisig'
+  feepayer?: string[]
```
Comment (Keith-CY) [inline packages/core/src/builtin-tasks/contract.ts:218]: `feepayer` could be `feePayer`

File: packages/core/src/builtin-tasks/contract.ts [hunk 4]
```diff
@@ -198,6 +210,181 @@ subtask('contract:deploy')
...
+  deployer?: string[]
+  feeRate?: number
+  export?: string
+  send: boolean
+}
+subtask('contract:upgrade')
+  .addParam('name', 'name of the contract to be deployed', '', paramTypes.string, true)
```
Comment (Keith-CY) [inline packages/core/src/builtin-tasks/contract.ts:225]:
```suggestion
  .addParam('name', 'name of the contract to upgrade', '', paramTypes.string, true)
```

File: packages/core/src/builtin-tasks/contract.ts [hunk 5]
```diff
@@ -198,6 +210,181 @@ subtask('contract:deploy')
...
+  .addParam('export', 'export transaction to file', '', paramTypes.path, true)
+  .addParam('send', 'send transaction directly', false, paramTypes.boolean, true)
+  .setAction(async (args: ContractUpgradeArgs, { config, run }) => {
+    const { name, feepayer, deployer, feeRate, signer, binPath, send = false, migrationPath } = args
+    const { ckbChain } = config
+
+    const feepayerInfos = feepayer ? parseFromInfoByCli(feepayer) : []
+    const deployerInfos = deployer ? parseFromInfoByCli(deployer) : []
+
+    const migrationFile: ContractMigration = await (async () => {
```
Comment (Keith-CY) [inline packages/core/src/builtin-tasks/contract.ts:253]: Which method makes it `async/await`

File: packages/core/src/builtin-tasks/contract.ts [hunk 6]
```diff
@@ -198,6 +210,181 @@ subtask('contract:deploy')
...
+                ),
+              ),
+              require_first_n: fromInfo.R,
+              threshold: fromInfo.M,
+            },
+          })
+        }
+      })
+
+      writeFileSync(exportPath, JSON.stringify(exportData, null, 2))
```
Comment (Keith-CY) [inline packages/core/src/builtin-tasks/contract.ts:377]: In what case will it be exported and sent simultaneously?

File: packages/core/src/contract.ts [hunk 7]
```diff
@@ -83,4 +83,59 @@ export class ContractDeployer {
...
+    send: () => Promise<Hash>
+  }> {
+    const { feeRate = 1000 } = options || {}
+    const contractBin = readFileSync(contractBinPath)
+
+    const typeIdScript = await (async () => {
+      const tx = await this.#rpc.getTransaction(contract.txHash)
+
+      return tx.transaction.outputs[contract.index].type!
+    })()
```
Comment (Keith-CY) [inline packages/core/src/contract.ts:113]: Why not run these methods directly but wrap them into an IIFE?

File: packages/core/src/builtin-tasks/contract.ts [hunk 8]
```diff
@@ -198,6 +210,181 @@ subtask('contract:deploy')
...
+
+    const conrtactBinPath = await (async () => {
+      // check bin path is absolute path
+      if (binPath && path.isAbsolute(binPath)) {
+        return binPath
+      }
+
+      if (binPath) {
+        return path.join(process.cwd(), binPath)
+      }
```
Comment (Daryl-L) [inline packages/core/src/builtin-tasks/contract.ts:293]: Isn't it better to coding like follow?

```typescript
      // check bin path is absolute path
      if (binPath) {
        if (binPath && path.isAbsolute(binPath)) {
          return binPath
        }

        return path.join(process.cwd(), binPath)
      }
```

File: packages/core/src/builtin-tasks/contract.ts [hunk 9]
```diff
@@ -198,6 +210,181 @@ subtask('contract:deploy')
...
+      deployerInfos[0],
+      feepayerInfos[0],
+      {
+        txHash: migration.tx_hash,
+        index: migration.index,
+        dataHash: migration.data_hash,
+      },
+      { feeRate },
+    )
+    const { tx } = deployTx
```
Comment (Daryl-L) [inline packages/core/src/builtin-tasks/contract.ts:342]:
```suggestion
    const { tx } = await contractDeployer.upgrade(
      conrtactBinPath,
      deployerInfos[0],
      feepayerInfos[0],
      {
        txHash: migration.tx_hash,
        index: migration.index,
        dataHash: migration.data_hash,
      },
      { feeRate },
    )
```

File: packages/core/src/builtin-tasks/contract.ts [hunk 10]
```diff
@@ -198,6 +210,181 @@ subtask('contract:deploy')
...
+  feeRate?: number
+  export?: string
+  send: boolean
+}
+subtask('contract:upgrade')
+  .addParam('name', 'name of the contract to be deployed', '', paramTypes.string, true)
+  .addParam('bin-path', 'path of contract bin file', '', paramTypes.path, true)
+  .addParam('migration-path', 'path of migration file')
+  .addParam('feepayer', 'address or multisig config of the transaction fee payer', '', paramTypes.string, false, true)
+  .addParam('deployer', 'address or multisig config of the contract deployer', '', paramTypes.string, false, true)
```
Comment (yanguoyu) [inline packages/core/src/builtin-tasks/contract.ts:229]: I can't get the point of `feepayer` and `deployer`, could we directly use the `deployer` cell to provide the tx fee?

File: packages/core/src/builtin-tasks/contract.ts (hunk 1 above)
Comment (PainterPuppets) [reply packages/core/src/builtin-tasks/contract.ts:22]: This migrations file is generated by `capsule`, and I'm not quite sure what `cell_recipes` means

**Review by yanguoyu — APPROVED** [review]

**Comment by Keith-CY** [comment]
Any update for the conversations?

**Review by Daryl-L — APPROVED** [review]

File: packages/core/src/builtin-tasks/contract.ts (hunk 6 above)
Comment (PainterPuppets) [reply packages/core/src/builtin-tasks/contract.ts:390]: Is this just an option for the user, or how about sending without exporting by default?

Comment (Keith-CY) [reply packages/core/src/builtin-tasks/contract.ts:390]: IMO, either `send` or `export` is enough because I cannot imagine a case of `send & export`

Comment (Keith-CY) [reply packages/core/src/builtin-tasks/contract.ts:390]: Any update on this conversation

Comment (PainterPuppets) [reply packages/core/src/builtin-tasks/contract.ts:390]: ok I'll update it to send by default in the absence of export

**Review by Keith-CY — APPROVED** [review]

**Pull request merged**